### PR TITLE
usbus/dfu: fix underflow condition while updating firmware

### PR DESCRIPTION
**Pull request by dylad**
### Contribution description

This PR fixes an underflow condition which can occurs if host sends less than 4 bytes of data during a DFU download operation.

While loading a new firmware through a `DFU_DNLOAD` request, DFU skips the `RIOTBOOT_FLASHWRITE_SKIPLEN` bytes as we don't need them in flash. This skip wasn't guarded so if the`DFU_DNLOAD` request contains less than `RIOTBOOT_FLASHWRITE_SKIPLEN` bytes, an underflow integer occurs with unexpected behaviour. Thus, add a proper check to prevent this situation from happening by stalling the request if it has less than `RIOTBOOT_FLASHWRITE_SKIPLEN` bytes of data.

edit:
The internal DFU state machine was also updated and will now returns dfuERROR in case of failure (underflow or flash write failure) to the host (through GET_STATUS request).
A new request handling was added to clear this error and return to dfuIDLE (CLRSTATUS request).
This way dfu-util safely abort the dfu upgrade procedure and indicate that something was wrong. A new dfu update can be made right after without restarting the device.

### Issues/PRs references
Reported-by szymonh through the forum, many thanks for his report !

### CODE REVIEW

File: sys/usb/usbus/dfu/dfu.c [hunk 1]
```diff
@@ -240,12 +255,16 @@ static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
...
     (void)state;
+
     usbus_dfu_device_t *dfu = (usbus_dfu_device_t *)handler;
     DEBUG("DFU: Request: 0x%x\n", setup->request);
 
     /* Process DFU class request */
     if (setup->type & USB_SETUP_REQUEST_TYPE_CLASS) {
-        _dfu_class_control_req(usbus, dfu, setup);
+        if (_dfu_class_control_req(usbus, dfu, setup) < 0) {
+            DEBUG("DFU: control request failed\n");
```
Comment (bergzand) [inline sys/usb/usbus/dfu/dfu.c:265]: Small nitpick

```suggestion
            DEBUG("DFU: control request failed: %u\n", setup->request);
```

Makes the debug line a bit more useful when debugging

Comment (dylad) [reply sys/usb/usbus/dfu/dfu.c]: Done

**Comment by bergzand** [comment]
Please squash!

**Review by bergzand — APPROVED** [review]
Tested this on an nrf52840dk with the `bootloaders/riotboot_dfu` example. Sending a 2 byte payload now properly returns an error to the host:

<details><summary> dfu-utils </summary>

```Shellsession
koen@morgen ~/dev/RIOT-review/bootloaders/riotboot_dfu $ dfu-util -D payload.bin  -a 1
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: File too short for DFU suffix
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 1209:7d02
Run-time device DFU version 0110
Claiming USB DFU Interface...
Setting Alternate Setting #1 ...
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 0110
Device returned transfer size 64
Copying data from PC to DFU device
Download        [                         ]   0%            0 bytesdfu-util: Error during download
koen@morgen ~/dev/RIOT-review/bootloaders/riotboot_dfu $ dfu-util -D payload.bin  -a 1
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: File too short for DFU suffix
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 1209:7d02
Run-time device DFU version 0110
Claiming USB DFU Interface...
Setting Alternate Setting #1 ...
Determining device status: state = dfuERROR, status = 0
dfuERROR, clearing status
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 0110
Device returned transfer size 64
Copying data from PC to DFU device
Download        [                         ]   0%            0 bytesdfu-util: Error during download
```

</details>

**Comment by bergzand** [comment]
Thanks for the quick fix @dylad

**Pull request merged**